source/mir/ndslice/slice.d

```diff
@@ -2544,6 +2544,51 @@ public:
         assert(imm == x);
         assert(mut == x);
     }
+    
+    size_t[N] numIndex(size_t n) @safe scope const
+    {
+        assert(n < elementCount, "numIndex: n must be less than elementCount");
+
+        size_t[N] output;
+        static if (N == 1) {
+            output[N - 1] = n;
+        } else {
+            output[0 .. (N - 1)] = _lengths[1 .. N];
+            static if (N > 2) {
+                foreach_reverse (i; Iota!(N - 2)) {
+                    output[i] *= output[i + 1];
+                }
+            }
+            size_t mul;
+            foreach (i; Iota!(N - 1)) {
+                mul = n / output[i];
+                n -= mul * output[i];
+                output[i] = mul;
+            }
+            output[N - 1] = n;
+        }
+        
+        return output;
+    }
+
+    size_t valStride(size_t n) @safe scope const
+    {
+        assert(n < elementCount, "valStride: n must be less than elementCount");
+
+        static if (kind == Contiguous || (kind == Canonical && N == 1)) {
+            return n;
+        } else static if (kind == Universal && N == 1) {
+            return n * _strides[0];
+        } else {
+            return indexStride(numIndex(n));
+        }
```

Comment:
I am not sure this compiles, but the code is almost the same that was used in FlattenedIterator.
```d
    ptrdiff_t valStride()(ptrdiff_t n) @safe scope const
    {
        ptrdiff_t _shift;
        foreach_reverse (i; Iota!(1, N))
        {
            immutable v = n / ptrdiff_t(_lengths[i]);
            n %= ptrdiff_t(_lengths[i]);
            static if (i == S)
                _shift += n;
            else
                _shift += n * _strides[i];
            n = v;
        }
        _shift += n * _strides[0];
        return _shift;
    }
```

Reply:
btw, any reason to make it public?

Reply:
@9il Not really. `indexStride` was public and I tried to take the same approach. I can make changes private.

Reply:
It is ok to make it public, but I am not sure about naming, and we need docs.

Reply:
Same as below. Happy to add more docs later today. Flexible on naming.

Reply:
BTW, this code is for Canonical and Universal slices. Contiguous should just return the `n`

Reply:
Thanks.